becquerel/parsers/n42.py

```diff
@@ -0,0 +1,222 @@
+    compression : str
+        compression: 'None' or 'CountedZeroes'.
+    """
+    text = text.strip().replace("\n", " ")
+    tokens = text.split()
+    data = [int(float(token)) for token in tokens]
+    if compression == "CountedZeroes":
+        new_data = []
+        k = 0
+        while k < len(data):
+            if data[k] != 0:
+                new_data.append(data[k])
+                k += 1
+            else:
+                new_data.extend([0] * data[k + 1])
+                k += 2
+        data = new_data
+    return np.array(data, dtype=int)
+
+
+@dataclass
+class N42RadMeasurement:
+    startime: str
+    realtime: float
+    livetime: float
+    counts: np.ndarray
+    # Corresponding name in the XML file, e.g. "EnergyCalibration-1"
+    # TODO: keep the becquerel Calibration?
+    calibration: str
+
+
+class N42File:
+    def __init__(self, path: Union[str, Path]) -> None:
+        self.path = path if isinstance(path, Path) else Path(path)
+        _ext = self.path.suffix.lower()
+        if _ext not in (".n42", ".xml"):
+            raise BecquerelParserError(f"File extension is incorrect: {_ext}")
+        # Open
+        self.tree = etree.parse(str(self.path))
+        self.root = self.tree.getroot()
+        # Validate
+        try:
+            _SCHEMA.validate(self.tree)
+        except Exception:
+            raise BecquerelParserError("TODO")
+        # N42 requires that this be the top level key
+        if self.root.tag != _ns("RadInstrumentData"):
+            raise BecquerelParserError(f"Invalid N42 root tag: {self.root.tag}")
+        # Read
+        self._parse()
+
+    def find_all(self, val: str, element: etree.Element = None) -> list:
+        """Find all element tags matching `val` and build a list of dicts from
+        `make_element_dict`.
+
+        Parameters
+        ----------
+        val : str
+            Search term
+        element : etree.Element, optional
+            lxml element to search within, by default None (uses the root element)
+
+        Returns
+        -------
+        list
+            Iterable of element dictionaries from `make_element_dict`.
+        """
+        element = self.root if element is None else element
+        return _find_all(val, element)
+
+    def _parse(self) -> None:
+        # Populate basic top level properties
+        self.rad_inst_info = self.find_all("RadInstrumentInformation")
+        self.rad_det_info = self.find_all("RadDetectorInformation")
+
+        # Grab energy calibrations
+        calibrations = self.find_all("EnergyCalibration")
+        self.calibrations = {}
+        for cal in calibrations:
+            txt = cal["CoefficientValues"]["value"]
+            coefs = np.array([float(x) for x in txt.split(" ")], dtype=float)
+            # TODO: are these always 3 params?
+            self.calibrations[cal["id"]] = Calibration(
+                "p[0] + p[1] * x + p[2] * x**2", coefs
+            )
```

Comment:
From the [n42 xml schema](https://www.nist.gov/programs-projects/documentary-standards-iec-62755-standard-n42-data-format):

> The values of the coefficients of a second-order polynomial describing the energy calibration in which the energies are expressed in keV.  The equation has the form:      E = T0 + T1*C + T2*C2      Where E is the energy in keV, C is the channel position, T0 is the offset coefficient, T1 is the gain coefficient, and T2 is the quadratic coefficient.  The first value in the list is term 0, the second term 1, and the third value term 2.  If the energy calibration is linear, the third coefficient (T2) shall be zero.